### PR TITLE
Implement "config set" subcommand

### LINKED ISSUE #531: jj should have a thing like `git config ...`
## Description

Learning to edit `toml` isn't going to be top of anyone's minds when they start using `jj`. A `jj config` would mean they don't need to

## Steps to Reproduce the Problem

1. See #530 

## Expected Behavior

`jj config user.name '...'`
`jj config user.email '...'`

## Actual Behavior

```sh
% jj
Config error: invalid TOML value, did you mean to use a quoted string? at line 2 column 8 in ../../../.jjconfig.toml
% jj
Config error: expected newline, found a plus at line 3 column 16 in ../../../.jjconfig.toml
```

## Specifications

- Platform:
- Version: 0.4.0

### PR DESCRIPTION
**Pull request by dbarnett**
Uses toml_edit to support simple config edits like:
  `jj config set --repo user.email "somebody@example.com"`

Fixes #531.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by dbarnett** [comment]
Hmm, my "cargo fmt" isn't wrapping these lines but the "Check formatting" insists the long line needs to be wrapped. Any idea why? I _just_ did rustup update.

```
-    // TODO(#531): Infer types based on schema, and support --type arg to explicitly specify.
+    // TODO(#531): Infer types based on schema, and support --type arg to explicitly
+    // specify.
```

**Comment by martinvonz** [comment]
> Hmm, my "cargo fmt" isn't wrapping these lines but the "Check formatting" insists the long line needs to be wrapped. Any idea why? I _just_ did rustup update.
> 
> ```
> -    // TODO(#531): Infer types based on schema, and support --type arg to explicitly specify.
> +    // TODO(#531): Infer types based on schema, and support --type arg to explicitly
> +    // specify.
> ```

We [use the nightly version in CI](https://github.com/martinvonz/jj/blob/c5e41a99c3b4aef43a88d864389685be394d7ff1/.github/workflows/build.yml#L59-L68) (I don't remember if there's a good reason for that), so try `cargo +nightly fmt` and see if that's the difference.

File: src/cli_util.rs [hunk 1]
```diff
@@ -1596,6 +1600,85 @@ pub fn serialize_config_value(value: &config::Value) -> String {
...
+    // TODO(#531): Prefer to create new structured tables vs inline.
+    let mut target_entry = doc.as_item_mut();
+    let mut key_parts_iter = key.split('.');
+    // Note: split guarantees at least one item.
+    let mut last_key_part = key_parts_iter.next().unwrap();
+    for key_part in &mut key_parts_iter {
+        target_entry = &mut target_entry[last_key_part];
+        last_key_part = key_part;
+    }
+    target_entry[last_key_part] = item;
```
Comment (yuja) [inline src/cli_util.rs:1645]: `item[str]` may fail if the `item` already exists and is not a table.
https://docs.rs/toml_edit/0.19.4/src/toml_edit/index.rs.html#52

I'm not sure what's the best way to look up nested tables, but maybe we'll have to use `.into_table()`, `Table::entry()`, etc.

File: src/cli_util.rs [hunk 2]
```diff
@@ -1596,6 +1600,85 @@ pub fn serialize_config_value(value: &config::Value) -> String {
...
+        | Ok(value @ toml_edit::Value::Integer(..))
+        | Ok(value @ toml_edit::Value::Float(..))
+        | Ok(value @ toml_edit::Value::String(..)) => toml_edit::value(value),
+        _ => toml_edit::value(value_str),
+    };
+    // TODO(#531): Prefer to create new structured tables vs inline.
+    let mut target_entry = doc.as_item_mut();
+    let mut key_parts_iter = key.split('.');
+    // Note: split guarantees at least one item.
+    let mut last_key_part = key_parts_iter.next().unwrap();
```
[the remark this answers is not in the document]
Comment (dbarnett) [reply src/cli_util.rs]: Nice! Done.

File: src/cli_util.rs (hunk 1 above)
Comment (dbarnett) [reply src/cli_util.rs]: Thanks! Code looks a little more awkward, but it works well (and fixes one of my TODOs to prefer structured tables vs. inline).

File: src/cli_util.rs [hunk 3]
```diff
@@ -1614,6 +1618,84 @@ pub fn serialize_config_value(value: &config::Value) -> String {
...
+        | Ok(value @ toml_edit::Value::Integer(..))
+        | Ok(value @ toml_edit::Value::Float(..))
+        | Ok(value @ toml_edit::Value::String(..)) => toml_edit::value(value),
+        _ => toml_edit::value(value_str),
+    };
+    let mut target_table = doc.as_table_mut();
+    let mut key_parts_iter = key.split('.');
+    // Note: split guarantees at least one item.
+    let last_key_part = key_parts_iter.next_back().unwrap();
+    for key_part in &mut key_parts_iter {
```
Comment (yuja) [inline src/cli_util.rs:1658]: nit: no need to borrow `key_parts_iter`

File: src/cli_util.rs [hunk 4]
```diff
@@ -1614,6 +1618,84 @@ pub fn serialize_config_value(value: &config::Value) -> String {
...
+        target_table = target_table
+            .entry(key_part)
+            .or_insert_with(|| toml_edit::Item::Table(toml_edit::Table::new()))
+            .as_table_mut()
+            .ok_or_else(|| {
+                user_error(format!(
+                    "Failed to set {key}: would overwrite non-table value with parent table"
```
Comment (yuja) [inline src/cli_util.rs:1665]: Can you add a test for overwrites?

File: src/cli_util.rs [hunk 5]
```diff
@@ -1614,6 +1618,84 @@ pub fn serialize_config_value(value: &config::Value) -> String {
...
+                ))
+            })?;
+    }
+    target_table[last_key_part] = item;
```
Comment (yuja) [inline src/cli_util.rs:1669]: We can also check that `target_table[last_key_part]` is not a table so `config set ui xxx` should fail.

(not required in this PR)

**Review by yuja — APPROVED** [review]
LGTM, thanks.

File: src/cli_util.rs (hunk 3 above)
Comment (dbarnett) [reply src/cli_util.rs]: Updated. (I was following https://rust-lang.github.io/rust-clippy/master/index.html#while_let_on_iterator where they suggested a mut borrow for some reason.)

File: src/cli_util.rs (hunk 5 above)
Comment (dbarnett) [reply src/cli_util.rs:1679]: Done, and added a test.

File: src/cli_util.rs (hunk 3 above)
Comment (yuja) [reply src/cli_util.rs]: Perhaps, clippy isn't clever enough to notice the iterator is never used thereafter.

**Pull request merged**